File: LeetCode/1728. Cat and Mouse II/solution.py
class Solution:
    def canMouseWin(self, grid: List[str], catJump: int, mouseJump: int) -> bool:
        n, m = len(grid), len(grid[0])
        def bfs(x,y):
            visited = set()
            visited.add((x,y))
            q = deque()
            q.append((x,y))
            while q:
                r, c = q.popleft()
                for dr, dc in [(0,-1),(0,1),(-1,0),(1,0)]:
                    dr += r
                    dc += c 
                    if 0 <= dr < n and 0 <= dc < m and (dr,dc) not in visited and grid[dr][dc] != '#':
                        visited.add((dr,dc))
                        q.append((dr,dc))
            return visited

        for i, row in enumerate(grid):
            for j, ch in enumerate(row):
                if ch == 'C':
                    catReach = bfs(i,j)
                    cat = (i,j)
                if ch == 'M':
                    mouseReach = bfs(i,j)
                    mouse = (i,j)
                if ch == 'F':
                    food = (i,j)
        # print(catReach,mouseReach,cat,mouse,food)
        
        def available(x,y,delta):
            yield x, y
            for dx, dy in [(0,-1),(0,1),(-1,0),(1,0)]:
                for k in range(1,delta+1):
                    xx, yy = x+dx*k, y+dy*k
                    if xx >= n or xx < 0 or yy >= m or yy < 0 or grid[xx][yy] == '#':
                        break
                    yield xx, yy
            return 0
        degree = Counter()
        for cat_x, cat_y in catReach:
            for mouse_x, mouse_y in mouseReach:
                for _, _ in available(cat_x,cat_y,catJump):
                    degree[cat_x,cat_y,mouse_x,mouse_y,True] += 1
                for _, _ in available(mouse_x,mouse_y,mouseJump):
                    degree[cat_x,cat_y,mouse_x,mouse_y,False] += 1
        pq = deque()
        seen = {}
        for x, y in catReach&mouseReach:
            seen[x,y,x,y,False] = False
            pq.append((x,y,x,y,False))
        for x, y in catReach:
            seen[x,y,food[0],food[1],True] = False
            pq.append((x,y,food[0],food[1],True))
        for x, y in mouseReach:
            seen[food[0],food[1],x,y,False] = False
            pq.append((food[0],food[1],x,y,False))

        # for x, y in catReach&mouseReach:
        #     if (x,y) == food:
        #         continue
        #     seen[x,y,x,y,False] = False
        #     seen[x,y,x,y,True] = True
        #     pq.append((x,y,x,y,False))
        #     pq.append((x,y,x,y,True))
        # for x, y in catReach:
        #     if (x,y) == food:
        #         continue
        #     seen[x,y,food[0],food[1],True] = False
        #     pq.append((x,y,food[0],food[1],True))
        # for x, y in mouseReach:
        #     if (x,y) == food:
        #         continue
        #     seen[food[0],food[1],x,y,False] = False
        #     pq.append((food[0],food[1],x,y,False))
        # print(seen)
        while pq:
            cat_x, cat_y, mouse_x, mouse_y, cat_turn = pq.popleft()
            win = seen[cat_x, cat_y, mouse_x, mouse_y, cat_turn]
            if (cat_x,cat_y) == cat and (mouse_x,mouse_y) == mouse and not cat_turn:
                return win 
            if cat_turn:
                for x, y in available(mouse_x,mouse_y,mouseJump):
                    degree[cat_x,cat_y,x,y,not cat_turn] -= 1
                    if (cat_x,cat_y,x,y,not cat_turn) in seen:
                        continue
                    if win and degree[cat_x,cat_y,x,y,not cat_turn] > 0:
                        continue
                    seen[cat_x,cat_y,x,y,not cat_turn] = not win
                    pq.append((cat_x,cat_y,x,y,not cat_turn))
            else:
                for x, y in available(cat_x,cat_y,catJump):
                    degree[x,y,mouse_x,mouse_y, not cat_turn] -= 1
                    if (x,y,mouse_x,mouse_y,not cat_turn) in seen:
                        continue
                    if win and degree[x,y,mouse_x,mouse_y,not cat_turn] > 0:
                        continue
                    seen[x,y,mouse_x,mouse_y,not cat_turn] = not win
                    pq.append((x,y,mouse_x,mouse_y,not cat_turn))
        return False

File: LeetCode/1728. Cat and Mouse II/solution.cpp
class Solution {
private:
    // f[cat_x][cat_y][mouse_x][mouse_y][is_cat_round] = 如果当前玩家必胜那么为 1，否则为 -1
    int f[8][8][8][8][2];
    // 统计每个状态的入度，用于拓扑排序
    int degree[8][8][8][8][2];

    static constexpr int dirs[4][2] = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

public:
    bool canMouseWin(vector<string>& grid, int catJump, int mouseJump) {
        int m = grid.size();
        int n = grid[0].size();

        auto findPosition = [&](char c) -> pair<int, int> {
            for (int i = 0; i < m; ++i) {
                for (int j = 0; j < n; ++j) {
                    if (grid[i][j] == c) {
                        return {i, j};
                    }
                }
            }
            return {-1, -1};
        };

        auto getNeighbors = [&](int x, int y, int bound) -> vector<pair<int, int>> {
            vector<pair<int, int>> ret = {{x, y}};
            for (int d = 0; d < 4; ++d) {
                int xx = x, yy = y;
                for (int _ = 1; _ <= bound; ++_) {
                    xx += dirs[d][0];
                    yy += dirs[d][1];
                    if (xx < 0 || xx >= m || yy < 0 || yy >= n || grid[xx][yy] == '#') {
                        break;
                    }
                    ret.emplace_back(xx, yy);
                }
            }
            return ret;
        };

        auto [cx, cy] = findPosition('C');
        auto [mx, my] = findPosition('M');
        auto [fx, fy] = findPosition('F');

        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (grid[i][j] != '#') {
                    for (int k = 0; k < m; ++k) {
                        for (int l = 0; l < n; ++l) {
                            if (grid[k][l] != '#') {
                                degree[i][j][k][l][0] = getNeighbors(k, l, mouseJump).size();
                                degree[i][j][k][l][1] = getNeighbors(i, j, catJump).size();
                            }
                        }
                    }
                }
            }
        }

        memset(f, 0, sizeof(f));
        queue<tuple<int, int, int, int, int>> q;

        // 猫和老鼠重合
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (grid[i][j] != '#' && grid[i][j] != 'F') {
                    f[i][j][i][j][0] = -1;
                    f[i][j][i][j][1] = 1;
                    q.emplace(i, j, i, j, 0);
                    q.emplace(i, j, i, j, 1);
                }
            }
        }

        // 猫 or 老鼠到达食物
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (grid[i][j] != '#' && grid[i][j] != 'F') {
                    f[fx][fy][i][j][0] = -1;
                    f[i][j][fx][fy][1] = -1;
                    q.emplace(fx, fy, i, j, 0);
                    q.emplace(i, j, fx, fy, 1);
                }
            }
        }

        while (!q.empty()) {
            auto [catx, caty, mousex, mousey, op] = q.front();
            q.pop();
            if (op == 1) {
                vector<pair<int, int>> neighbors = getNeighbors(mousex, mousey, mouseJump);
                for (auto [x, y]: neighbors) {
                    --degree[catx][caty][x][y][op ^ 1];
                    if (!f[catx][caty][x][y][op ^ 1]) {
                        if (f[catx][caty][mousex][mousey][op] == -1) {
                            f[catx][caty][x][y][op ^ 1] = 1;
                            q.emplace(catx, caty, x, y, op ^ 1);
                        }
                        else if (degree[catx][caty][x][y][op ^ 1] == 0) {
                            f[catx][caty][x][y][op ^ 1] = -1;
                            q.emplace(catx, caty, x, y, op ^ 1);
                        }
                    }
                }
            }
            else {
                vector<pair<int, int>> neighbors = getNeighbors(catx, caty, catJump);
                for (auto [x, y]: neighbors) {
                    --degree[x][y][mousex][mousey][op ^ 1];
                    if (!f[x][y][mousex][mousey][op ^ 1]) { 
                        if (f[catx][caty][mousex][mousey][op] == -1) {
                            f[x][y][mousex][mousey][op ^ 1] = 1;
                            q.emplace(x, y, mousex, mousey, op ^ 1);
                        }
                        else if (degree[x][y][mousex][mousey][op ^ 1] == 0) {
                            f[x][y][mousex][mousey][op ^ 1] = -1;
                            q.emplace(x, y, mousex, mousey, op ^ 1);
                        }
                    }
                }
            }
        }

        return f[cx][cy][mx][my][0] == 1;
    }
};

